### PR TITLE
Used functional all-reduce for amax reduction

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #217
* #216
* #215
* #214
* __->__ #219
* #220

Dynamo cannot remap the in-place max all-reduce to its functional equivalent on PyTorch `main` branch.
https://github.com/pytorch/pytorch/issues/120082

### CODE REVIEW

**Review by drisspg — APPROVED** [review]
Looks great!

**Comment by awgu** [comment]
I will hold off on landing this since @yifuwang will land the Dynamo rewrite fix soon.

File: float8_experimental/float8_utils.py [hunk 1]
```diff
@@ -75,7 +75,13 @@ def tensor_to_amax(x, distributed_reduction=False):
     # If the user did not ask for it, assume that it will
     # happen elsewhere.
     if distributed_reduction and dist.is_initialized():
-        dist.all_reduce(amax, op=dist.ReduceOp.MAX)
+        # TODO: Dynamo rewriting synchronous in-place collectives fails:
+        # https://github.com/pytorch/pytorch/issues/120082
+        # Use functional all-reduce to avoid graph breaking.
+        amax = dist._functional_collectives.all_reduce(
+            amax, "MAX", list(range(dist.get_world_size()))
```
Comment (yifuwang) [inline float8_experimental/float8_utils.py:82]: Ranks + tag as the process group identifier has been deprecated. Can we pass `dist.group.WORLD` or `dist.group.WORLD.group_name` here?

Comment (awgu) [reply float8_experimental/float8_utils.py:82]: If we are not landing this PR, then is it okay to leave the call as `dist.all_reduce(amax, op=dist.ReduceOp.MAX)` and wait for your Dynamo rewrite changes?

Comment (yifuwang) [reply float8_experimental/float8_utils.py:82]: Oh the changes you are referring to are for rewriting functional collective. The options I mentioned above should already work :)

Let me know if it doesn't though.

**Comment by yifuwang** [comment]
> I will hold off on landing this since @yifuwang will land the Dynamo rewrite fix soon.

Thanks @awgu! The on-going PRs are for rewriting non-functional collectives. If you are already using functional collectives, I don't think (hopefully :) ) there are any blockers

**Pull request closed without merging**